core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/jdbc/JDBCUserStoreManager.java

```diff
@@ -4882,4 +4928,112 @@ private boolean isH2DB(Connection dbConnection) throws Exception {
 
         return H2.equalsIgnoreCase(DatabaseCreator.getDatabaseType(dbConnection));
     }
+
+    /**
+     * Circuit Breaker pattern added for DB connection retrieval
+     *          Renamed original getDBConnection() method to getConnection() to minimize changes
+     *
+     * @return DB Connection
+     * @throws UserStoreException if unknown occurred while getting database connection.
+     * @throws SQLException if error occurred while retrieving database connection
+     */
+    protected Connection getDBConnection() throws UserStoreException, SQLException {
+
+        Connection dbConnection = null;
+
+        switch (jdbcConnectionCircuitBreakerState) {
+            case CIRCUIT_STATE_OPEN:
+                long circuitOpenDuration = System.currentTimeMillis() - thresholdStartTime;
+                if (circuitOpenDuration >= thresholdTimeoutInMilliseconds) {
+                    try {
+                        if (log.isDebugEnabled()) {
+                            log.debug("Trying to obtain JDBC connection for " + getMyDomainName() +" domain" +
+                                    " when circuit breaker state: "
+                                    + jdbcConnectionCircuitBreakerState + " and circuit breaker open duration: "
+                                    + circuitOpenDuration + "ms.");
+                        }
+                        dbConnection = getConnection();
+                        jdbcConnectionCircuitBreakerState = CIRCUIT_STATE_CLOSE;
+                        thresholdStartTime = 0;
+                        break;
+                    } catch (Exception e) {
+                        log.error("Error occurred while obtaining connection", e);
+                        thresholdStartTime = System.currentTimeMillis();
+                        if (log.isDebugEnabled()) {
+                            log.debug("JDBC connection circuit breaker state set to: "
+                                    + jdbcConnectionCircuitBreakerState);
+                        }
+                        throw new CircuitBreakerException("Error occurred while obtaining connection.", e);
+                    }
+                } else {
+                    throw new CircuitBreakerException(
+                            "JDBC Connection circuit breaker is in open state for " + circuitOpenDuration
+                                    + "ms and has not reach the threshold timeout: " + thresholdTimeoutInMilliseconds
+                                    + "ms, hence avoid establishing the connection for " + getMyDomainName() +
+                                    " domain");
+                }
```

Comment:
can we extract the switch cases logics to private methods

Reply:
These aren't reusable and are different from each condition, therefore moving it to different methods seem redundant. What do you think?

Reply:
It improves the readability

Reply:
Agreed and moved the logic to private methods as requested.